// @open-dy/node-server-sdk/dist/esm/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dySDK = void 0;
var dycloud_db_1 = require("@open-dy/dycloud-db");
var axios_1 = __importDefault(require("axios"));
var https = __importStar(require("https"));
var reportHttp = axios_1.default.create({ timeout: 60 });
var databaseHttp = axios_1.default.create({ timeout: 60000 });
var Request = (function () {
    function Request() {
    }
    Request.prototype.report = function (url, params, time, httpCode, statusCode) {
        return __awaiter(this, void 0, void 0, function () {
            var randomNum, reportParams, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        randomNum = Math.floor(Math.random() * 100) + 1;
                        if (!(randomNum <= 10)) return [3, 4];
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        reportParams = {
                            counter_value: 1,
                            timer_value: time,
                            tag: {
                                domain: url,
                                http_code: String(httpCode),
                                status_code: String(statusCode),
                                action: params === null || params === void 0 ? void 0 : params.action,
                            },
                        };
                        return [4, reportHttp.post('http://metrics.dyc.ivolces.com', reportParams, {
                                headers: {
                                    'content-type': 'application/json',
                                },
                            })];
                    case 2:
                        _b.sent();
                        return [3, 4];
                    case 3:
                        _a = _b.sent();
                        return [3, 4];
                    case 4: return [2];
                }
            });
        });
    };
    Request.prototype.send = function (action, data, _) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        return __awaiter(this, void 0, void 0, function () {
            var params, startTime, url, res, error_1;
            return __generator(this, function (_m) {
                switch (_m.label) {
                    case 0:
                        params = Object.assign({}, data, {
                            action: action,
                        });
                        startTime = Date.now();
                        url = process.env.DYC_CLOUD_DATABASE_URL
                            ? process.env.DYC_CLOUD_DATABASE_URL
                            : 'http://cloud-db.dyc.ivolces.com/api/cloud_database/exec_cloud_database_cmd';
                        _m.label = 1;
                    case 1:
                        _m.trys.push([1, 3, , 4]);
                        return [4, databaseHttp.post(url, params, {
                                headers: {
                                    'content-type': 'application/json',
                                },
                            })];
                    case 2:
                        res = _m.sent();
                        if ((_a = res === null || res === void 0 ? void 0 : res.data) === null || _a === void 0 ? void 0 : _a.status_code) {
                            void this.report(url, params, Date.now() - startTime, 200, res.data.status_code);
                            throw new dycloud_db_1.DataBaseError({
                                errMsg: "".concat(res.data.status_message, " requestId: ").concat((_d = (_c = (_b = res.data) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.request_id) !== null && _d !== void 0 ? _d : ''),
                                errorCode: res.data.status_code,
                            });
                        }
                        void this.report(url, params, Date.now() - startTime, 200, 0);
                        return [2, ((_e = res === null || res === void 0 ? void 0 : res.data) === null || _e === void 0 ? void 0 : _e.data) || {}];
                    case 3:
                        error_1 = _m.sent();
                        void this.report(url, params, Date.now() - startTime, (_f = error_1 === null || error_1 === void 0 ? void 0 : error_1.response) === null || _f === void 0 ? void 0 : _f.status, 0);
                        throw new dycloud_db_1.DataBaseError({
                            errMsg: "".concat((_g = error_1 === null || error_1 === void 0 ? void 0 : error_1.response) === null || _g === void 0 ? void 0 : _g.statusText, " requestId: ").concat((_k = (_j = (_h = error_1 === null || error_1 === void 0 ? void 0 : error_1.response) === null || _h === void 0 ? void 0 : _h.headers) === null || _j === void 0 ? void 0 : _j['x-tt-logid']) !== null && _k !== void 0 ? _k : ''),
                            errorCode: (_l = error_1 === null || error_1 === void 0 ? void 0 : error_1.response) === null || _l === void 0 ? void 0 : _l.status,
                        });
                    case 4: return [2];
                }
            });
        });
    };
    return Request;
}());
var ServiceContext = (function () {
    function ServiceContext(context) {
        this.context = context;
    }
    ServiceContext.prototype.getContext = function () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
        return {
            source: (_b = (_a = this === null || this === void 0 ? void 0 : this.context) === null || _a === void 0 ? void 0 : _a.headers) === null || _b === void 0 ? void 0 : _b['x-tt-source'],
            ip: (_d = (_c = this === null || this === void 0 ? void 0 : this.context) === null || _c === void 0 ? void 0 : _c.headers) === null || _d === void 0 ? void 0 : _d['x-forwarded-for'],
            appId: (_e = this.context.headers) === null || _e === void 0 ? void 0 : _e['x-tt-appid'],
            envId: (_g = (_f = this === null || this === void 0 ? void 0 : this.context) === null || _f === void 0 ? void 0 : _f.headers) === null || _g === void 0 ? void 0 : _g['x-tt-envid'],
            openId: (_j = (_h = this === null || this === void 0 ? void 0 : this.context) === null || _h === void 0 ? void 0 : _h.headers) === null || _j === void 0 ? void 0 : _j['x-tt-openid'],
            unionId: (_l = (_k = this === null || this === void 0 ? void 0 : this.context) === null || _k === void 0 ? void 0 : _k.headers) === null || _l === void 0 ? void 0 : _l['x-tt-unionid'],
            anonymousOpenid: (_o = (_m = this === null || this === void 0 ? void 0 : this.context) === null || _m === void 0 ? void 0 : _m.headers) === null || _o === void 0 ? void 0 : _o['x-tt-anonymous-openid'],
            roomId: (_q = (_p = this === null || this === void 0 ? void 0 : this.context) === null || _p === void 0 ? void 0 : _p.headers) === null || _q === void 0 ? void 0 : _q['x-room-id'],
            anchorOpenId: (_s = (_r = this === null || this === void 0 ? void 0 : this.context) === null || _r === void 0 ? void 0 : _r.headers) === null || _s === void 0 ? void 0 : _s['x-anchor-openid'],
            avatarUrl: (_u = (_t = this === null || this === void 0 ? void 0 : this.context) === null || _t === void 0 ? void 0 : _t.headers) === null || _u === void 0 ? void 0 : _u['x-avatar-url'],
            nickName: (_w = (_v = this === null || this === void 0 ? void 0 : this.context) === null || _v === void 0 ? void 0 : _v.headers) === null || _w === void 0 ? void 0 : _w['x-nick-name'],
        };
    };
    ServiceContext.prototype.callContainer = function (params) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var selfServiceId;
            return __generator(this, function (_b) {
                selfServiceId = (_a = this === null || this === void 0 ? void 0 : this.context) === null || _a === void 0 ? void 0 : _a.headers['x-tt-serviceid'];
                return [2, new Promise(function (resolve, reject) {
                        axios_1.default
                            .request({
                            transformResponse: params === null || params === void 0 ? void 0 : params.transformResponse,
                            method: params.method,
                            url: "http://".concat(selfServiceId, "-").concat(params.serviceId, ".inner-call.dyc.ivolces.com").concat(params.path),
                            data: params === null || params === void 0 ? void 0 : params.data,
                            headers: params === null || params === void 0 ? void 0 : params.headers,
                            params: params === null || params === void 0 ? void 0 : params.querys,
                        })
                            .then(function (response) {
                            resolve(response.data);
                        })
                            .catch(function (error) {
                            reject(error);
                        });
                    })];
            });
        });
    };
    ServiceContext.prototype.openApiInvoke = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, new Promise(function (resolve, reject) {
                        var httpsAgent = new https.Agent({
                            rejectUnauthorized: false,
                        });
                        axios_1.default.defaults.httpsAgent = httpsAgent;
                        axios_1.default
                            .request({
                            transformResponse: params === null || params === void 0 ? void 0 : params.transformResponse,
                            method: params.method,
                            url: params.url,
                            data: params === null || params === void 0 ? void 0 : params.data,
                            headers: params === null || params === void 0 ? void 0 : params.headers,
                            params: params === null || params === void 0 ? void 0 : params.querys,
                        })
                            .then(function (response) {
                            resolve(response.data);
                        })
                            .catch(function (error) {
                            reject(error);
                        });
                    })];
            });
        });
    };
    return ServiceContext;
}());
exports.dySDK = (function () {
    var instance;
    return {
        database: function () {
            if (instance === undefined) {
                instance = new dycloud_db_1.Db(Request);
            }
            return instance;
        },
        context: function (context) {
            return new ServiceContext(context);
        },
    };
})();
